File: 655. Print Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<String>> printTree(TreeNode root) {
        int[] values = find_depth_width(root);
        int depth = values[0];
        int width = values[1];
        List<List<String>> ret = new ArrayList<>();
        for(int row = 0; row < depth; row++){
            ret.add(new ArrayList<>());
            for(int col = 0; col < width; col++){
                ret.get(row).add("");
            }
        }
        helper(ret, 0, 0, width-1, root);
        return ret;
    }

    public int[] find_depth_width(TreeNode node){
        if(node == null){
            return new int[]{0, 0};
        }
        if(node.left == null && node.right == null){
            return new int[]{1, 1};
        }
        int[] left = find_depth_width(node.left);
        int[] right = find_depth_width(node.right);
        return new int[]{Math.max(left[0], right[0]) + 1, 2 * Math.max(left[1], right[1]) + 1};
    }

    public void helper(List<List<String>> ans, int depth, int left, int right, TreeNode root){
        int mid = left + (right - left) / 2;
        ans.get(depth).set(mid, root.val + "");
        if(root.left != null){
            helper(ans, depth + 1, left, mid-1, root.left);
        }
        if(root.right != null){
            helper(ans, depth+1, mid+1, right, root.right);
        }
    }
}

File: 655. Print Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def printTree(self, root: TreeNode) -> List[List[str]]:
        
        def helper(root):
            if not root:
                return 0, 0
            if not root.left and not root.right:
                return 1, 1
            left_depth, left_width = helper(root.left)
            right_depth, right_width = helper(root.right)
            return max(left_depth, right_depth)+1, 2 * max(left_width, right_width) + 1
        row, column = helper(root)
        ret = [[""] * column for _ in range(row)]
        
        def helper2(node, left, right, depth):
            mid = left + (right - left) // 2
            ret[depth][mid] = str(node.val)
            if node.left:
                helper2(node.left, left, mid-1, depth+1)
            if node.right:
                helper2(node.right, mid+1, right, depth+1)
        helper2(root, 0, column-1, 0)
        return ret
            